src/v2/coins/abstractUtxoCoin.ts

```diff
@@ -833,46 +826,56 @@ class AbstractUtxoCoin extends BaseCoin {
+      // is native or not. The inputClassification is determined based on whether or not the input is native to give an
+      // accurate classification. Note that p2shP2wsh inputs will be classified as p2sh and not p2wsh.
       decompiledSigScript = currentInput.witness;
-      signatureScript = bitcoin.script.compile(decompiledSigScript);
-      if (isNativeSegwit) {
-        const lastWitness = _.last(transaction.ins[inputIndex].witness);
-        // const inputScriptHash = bitcoin.crypto.hash160(lastWitness);
-        const witnessScriptHash = bitcoin.crypto.sha256(lastWitness);
-        const prevOutScript = bitcoin.script.witnessScriptHash.output.encode(witnessScriptHash);
-        // we are faking a signature script for the verification
-        signatureScript = bitcoin.script.compile([prevOutScript]);
+      if (isNativeSegwitInput) {
+        inputClassification = bitcoin.script.classifyWitness(bitcoin.script.compile(decompiledSigScript), true);
+      } else {
+        inputClassification = bitcoin.script.classifyInput(currentInput.script, true);
       }
-    }
+    } else {
+      inputClassification = bitcoin.script.classifyInput(currentInput.script, true);
+      decompiledSigScript = bitcoin.script.decompile(currentInput.script);
 
-    const inputClassification = bitcoin.script.classifyInput(signatureScript, true);
-    if (inputClassification !== bitcoin.script.types.P2SH) {
-      if (inputClassification === bitcoin.script.types.P2PKH) {
-        const [signature, publicKey] = decompiledSigScript;
-        const publicKeys = [publicKey];
-        const signatures = [signature];
+    }
 
-        const pubScript = bitcoin.script.pubKeyHash.output.encode(bitcoin.crypto.hash160(publicKey));
-        return { signatures, publicKeys, isSegwitInput, inputClassification, pubScript: pubScript };
-      }
+    if (inputClassification === bitcoin.script.types.P2PKH) {
+      const [signature, publicKey] = decompiledSigScript;
+      const publicKeys = [publicKey];
+      const signatures = [signature];
+      const pubScript = bitcoin.script.pubKeyHash.output.encode(bitcoin.crypto.hash160(publicKey));
+
+      return { signatures, publicKeys, isSegwitInput, inputClassification, pubScript };
+    } else if (inputClassification === bitcoin.script.types.P2SH
+        || inputClassification === bitcoin.script.types.P2WSH) {
+      // Note the assumption here that if we have a p2sh or p2wsh input it will be multisig (appropriate because the
+      // BitGo platform only supports multisig within these types of inputs). Signatures are all but the last entry in
+      // the decompiledSigScript. The redeemScript/witnessScript (depending on which type of input this is) is the last
+      // entry in the decompiledSigScript (denoted here as the pubScript). The public keys are the second through
+      // antepenultimate entries in the decompiledPubScript. See below for a visual representation of the typical 2-of-3
```

Comment:
> antepenultimate

beautiful

Reply:
i cannot take credit for this. It was present before this change. :grimacing: